core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala

```diff
@@ -596,6 +596,95 @@ class DeltaSuite extends QueryTest
     }
   }
 
+  test("batch write: append, dynamic partition overwrite") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(1, 2, 3).toDF
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(1, 5).toDF
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.toDF.select($"value".as[Int]), 1, 2, 5)
+
+      Seq(("a", "x"), ("b", "y"), ("c", "x")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("overwriteSchema", true)
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("a", "x"), ("d", "x")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.toDF.select($"value".as[String]), "a", "b", "d")
+    }
+  }
+
+  test("batch write: append, dynamic partition overwrite with replaceWhere") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq((1, "x"), (2, "y"), (3, "z")).toDF("value", "sub")
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part", "sub")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq((3, "x"), (5, "x")).toDF("value", "sub")
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part", "sub")
+        .mode("overwrite")
+        .option(DeltaOptions.REPLACE_WHERE_OPTION, "part = 1")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.toDF.select($"value".as[Int], $"sub".as[String]),
+        (2, "y"), (3, "x"), (5, "x"))
+    }
+  }
+
+  test("batch write: append, overwrite without partitions should ignore partition overwrite mode") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq (1, 2, 3).toDF
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
```

Comment:
What should the behavior here if the initial data is partitioned, but the new data is not? cc: @tdas 

I checked using parquet, and when writing unpartitioned data to a partitioned table with `partitionOverwriteMode=dynamic` it overwrites all partitions.

Right now, this PR does not do that, would need...
```
val useDynamicPartitionOverwriteMode =
      partitionColumns && options.isDynamicPartitionOverwriteMode
```